pkg/raft/log.go

```diff
@@ -23,6 +23,28 @@ import (
 	pb "github.com/cockroachdb/cockroach/pkg/raft/raftpb"
 )
 
+// LogSnapshot encapsulates a point-in-time state of the raft log accessible
+// outside the raft package for reads.
+//
+// To access it safely, the user must not mutate the underlying raft log storage
+// between the snapshot is obtained and the reads are done.
+//
+// TODO(pav-kv): this should be part of the Ready API. Instead of pre-fetching
+// entries (e.g. the committed entries subject to state machine application),
+// allow the application to read them from LogSnapshot in the Ready handler.
+// This gives the application direct control on resource allocation, and
+// flexibility to do raft log IO without blocking RawNode operation.
+type LogSnapshot struct {
+	// first is the first available log index.
+	first uint64
+	// storage contains the stable log entries.
+	storage Storage
```

Comment:
You probably saw my comment on the prototype. `LogSnapshot` is definitely a step in the right direction, so I don't think it would make sense to try to take `Storage` out of it at this point. In other words, ignore that comment as far as making code changes now goes. In spirit, I still think there's something there, but it may be for another day (which may never come).